src-self-hosted/link.zig

```diff
@@ -89,447 +115,570 @@ pub fn openBinFile(allocator: *Allocator, file: fs.File, options: Options) !ElfF
-
-    /// A list of text blocks that have surplus capacity. This list can have false
-    /// positives, as functions grow and shrink over time, only sometimes being added
-    /// or removed from the freelist.
-    ///
-    /// A text block has surplus capacity when its overcapacity value is greater than
-    /// minimum_text_block_size * alloc_num / alloc_den. That is, when it has so
-    /// much extra capacity, that we could fit a small new symbol in it, itself with
-    /// ideal_capacity or more.
-    ///
-    /// Ideal capacity is defined by size * alloc_num / alloc_den.
-    ///
-    /// Overcapacity is measured by actual_capacity - ideal_capacity. Note that
-    /// overcapacity can be negative. A simple way to have negative overcapacity is to
-    /// allocate a fresh text block, which will have ideal capacity, and then grow it
-    /// by 1 byte. It will then have -1 overcapacity.
-    text_block_free_list: std.ArrayListUnmanaged(*TextBlock) = std.ArrayListUnmanaged(*TextBlock){},
-    last_text_block: ?*TextBlock = null,
-
-    /// `alloc_num / alloc_den` is the factor of padding when allocating.
-    const alloc_num = 4;
-    const alloc_den = 3;
-
-    /// In order for a slice of bytes to be considered eligible to keep metadata pointing at
-    /// it as a possible place to put new symbols, it must have enough room for this many bytes
-    /// (plus extra for reserved capacity).
-    const minimum_text_block_size = 64;
-    const min_text_capacity = minimum_text_block_size * alloc_num / alloc_den;
+pub const File = struct {
+    tag: Tag,
+    pub fn cast(base: *File, comptime T: type) ?*T {
+        if (base.tag != T.base_tag)
+            return null;
 
-    pub const ErrorFlags = struct {
-        no_entry_point_found: bool = false,
-    };
+        return @fieldParentPtr(T, "base", base);
+    }
 
-    pub const TextBlock = struct {
-        /// Each decl always gets a local symbol with the fully qualified name.
-        /// The vaddr and size are found here directly.
-        /// The file offset is found by computing the vaddr offset from the section vaddr
-        /// the symbol references, and adding that to the file offset of the section.
-        /// If this field is 0, it means the codegen size = 0 and there is no symbol or
-        /// offset table entry.
-        local_sym_index: u32,
-        /// This field is undefined for symbols with size = 0.
-        offset_table_index: u32,
-        /// Points to the previous and next neighbors, based on the `text_offset`.
-        /// This can be used to find, for example, the capacity of this `TextBlock`.
-        prev: ?*TextBlock,
-        next: ?*TextBlock,
-
-        pub const empty = TextBlock{
-            .local_sym_index = 0,
-            .offset_table_index = undefined,
-            .prev = null,
-            .next = null,
+    pub fn makeWritable(base: *File, dir: fs.Dir, sub_path: []const u8) !void {
+        try switch (base.tag) {
```

Comment:
Better to remove `try` here and instead put `return` statements in each prong. In the case of error return traces, it helps to point to the prong that was taken.

Reply:
Gotcha, good to know